apm-agent-plugins/apm-asynchttpclient-plugin/src/main/java/co/elastic/apm/agent/asynchttpclient/AbstractAsyncHttpClientInstrumentation.java

```diff
@@ -0,0 +1,180 @@
+    }
+
+    @Override
+    public ElementMatcher.Junction<ClassLoader> getClassLoaderMatcher() {
+        return not(isBootstrapClassLoader())
+            .and(classLoaderCanLoadClass("org.asynchttpclient.AsyncHandler"));
+    }
+
+    public static class AsyncHttpClientInstrumentation extends AbstractAsyncHttpClientInstrumentation {
+
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        private static void onBeforeExecute(@Advice.Argument(value = 0) Request request,
+                                            @Advice.Argument(value = 1) AsyncHandler<?> asyncHandler,
+                                            @Advice.Local("span") Span span) {
+            if (tracer == null || tracer.getActive() == null) {
+                return;
+            }
+
+            final TraceContextHolder<?> parent = tracer.getActive();
+            span = HttpClientHelper.startHttpClientSpan(parent, request.getMethod(), request.getUri().toUrl(), request.getUri().getHost());
+
+            if (span != null) {
+                span.activate();
+                request.getHeaders().add(TraceContext.TRACE_PARENT_HEADER, span.getTraceContext().getOutgoingTraceParentHeader().toString());
+                handlerSpanMap.put(asyncHandler, span);
+            }
+        }
+
+        @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class)
+        private static void onAfterExecute(@Advice.Local("span") @Nullable Span span,
+                                           @Advice.Argument(value = 1) AsyncHandler<?> asyncHandler,
+                                           @Advice.Thrown @Nullable Throwable t) {
+            if (span != null) {
+                span.deactivate();
+                if (t != null) {
+                    handlerSpanMap.remove(asyncHandler);
+                    span.captureException(t).end();
+                }
+            }
+        }
+
+        @Override
+        public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+            return named("org.asynchttpclient.DefaultAsyncHttpClient");
+        }
+
+        @Override
+        public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+            return named("executeRequest")
+                .and(takesArgument(0, named("org.asynchttpclient.Request")))
+                .and(takesArgument(1, named("org.asynchttpclient.AsyncHandler")));
+        }
+    }
+
+    public abstract static class AbstractAsyncHandlerInstrumentation extends AbstractAsyncHttpClientInstrumentation {
+
+        private final ElementMatcher<? super MethodDescription> methodMatcher;
+
+        protected AbstractAsyncHandlerInstrumentation(ElementMatcher<? super MethodDescription> methodMatcher) {
+            this.methodMatcher = methodMatcher;
+        }
+
+        @Override
+        public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+            return hasSuperType(named("org.asynchttpclient.AsyncHandler"));
```

Comment:
Probably quite a costly matcher. Maybe we should only look within `org.asynchttpclient` and `application_packages` if defined.

Reply:
Costly even with the provided `getClassLoaderMatcher()` implementation?

Reply:
That certainly makes it less severe. Only if there are lots of classes in (child) classloaders which load `asynchttpclient` matching will be slow.

Reply:
You mean like if it is on the main app server lib for example?
Well, if you are worried about that, your suggestion seems efficient and good... :)

Reply:
I'm a bit worried that there are frameworks which use asynchttpclient under the hood and which define their own `AsyncHandlers`. We wouldn't match those then. I'd say let's continue with how it is now and optimize if we see this becomes a problem.